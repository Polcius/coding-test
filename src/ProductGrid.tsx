import React, { FunctionComponent } from "react";
import styled from "styled-components";
import Card from "./Card";
import mockData from "./products.json";

const GridWrapper = styled.section`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
  grid-gap: 1rem;
`;

type ProductGridProps = {
  sizeFilter: string;
};

const ProductGrid: FunctionComponent<ProductGridProps> = ({ sizeFilter }) => {
  const filteredItems = mockData.filter(item => item.size.includes(sizeFilter));
  return (
    <GridWrapper>
      {filteredItems.map(product => (
        <Card
          key={product.index}
          image={product.productImage}
          name={product.productName}
          price={product.price}
          isSale={product.isSale}
          isExclusive={product.isExclusive}
        />
      ))}
    </GridWrapper>
  );
};

export default ProductGrid;
