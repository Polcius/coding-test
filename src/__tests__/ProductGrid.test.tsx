import React from "react";
import { render, cleanup, waitForElement } from "@testing-library/react";
import ProductGrid from "../ProductGrid";
import mockData from "../products.json";

global.fetch = require("jest-fetch-mock");
afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();

test("<ProductGrid />", async () => {
  fetch.mockResponseOnce(JSON.stringify(mockData));

  const { getByTestId, getAllByTestId, queryByTestId } = render(
    <ProductGrid sizeFilter="" />
  );

  await waitForElement(() => getByTestId("wrapper"));
  expect(queryByTestId("loading")).toBeFalsy();
  expect(getAllByTestId("product").length).toBe(mockData.length);
});

test("<ProductGrid /> API fail", async () => {
  fetch.mockReject(new Error("fake error message"));

  const { getByTestId, queryByTestId } = render(<ProductGrid sizeFilter="" />);
  expect(getByTestId("loading")).toBeTruthy();
  await waitForElement(() => getByTestId("wrapper"));
  expect(queryByTestId("loading")).toBeFalsy();
  expect(queryByTestId("error")).toBeTruthy();
});
