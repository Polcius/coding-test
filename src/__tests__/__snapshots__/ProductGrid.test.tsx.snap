// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductGrid /> API fail 1`] = `
<div
  data-testid="error"
>
  Something went wrong...
</div>
`;
